test/puppetlabs/pcp/messaging_test.clj

```diff
@@ -72,6 +71,41 @@
   [cn handler-fn]
   (connect-client-config (client-config cn) handler-fn))
 
+(defmacro eventually-logged?
+  [logger-id level pred & body]
+  `(with-log-level ~logger-id ~level
+     (with-log-suppressed-unless-notable (constantly false)
+       (let [log# (atom [])]
+         (with-logging-to-atom ~logger-id log#
+           (do ~@body)
+           (loop [n# 15] ;; loop up to 15 times to see if predicate is satisfied
+             (let [log-text# (map (fn [x#] (.getMessage x#)) (deref log#))]
+               (cond
+                 (neg? n#) (is (some ~pred log-text#))
+                 (some ~pred log-text#) (is (some ~pred log-text#))
+                 :else (do (Thread/sleep 100)
+                           (recur (dec n#)))))))))))
+
+(defn call-with-retries
+  [f]
+  (loop [n 10]
+    (let [result (f)]
+      (if (or result (neg? n))
+        result
+        (do (Thread/sleep 100)
+            (recur (dec n)))))))
+
+(defmacro is-eventually?
+  [& body]
+  `(is (loop [n# 10]
+         (let [result# (do ~@body)]
+           (if (or (= 0 (:fail result#)) (neg? n#))
+             result#
+             (do (Thread/sleep 100)
+                 (recur (dec n#))))))))
+
+;(clojure.pprint/pprint (macroexpand '(is-eventually? (logged? 10 10))))
```

Comment:
It'd be nice to remove this line, and put the utilities in a different file. Maybe worth moving to kitchensink someday.